File: src/services/auth/auth.service.ts
import { prisma } from "@lib/prisma.js";
import { StatusCodes } from "http-status-codes";
import { ApiError } from "@errors/ApiError.js";
import { JWTUtils } from "@utils/jwt.utils.js";
import { PasswordUtils } from "@utils/password.utils.js";
import type { LoginResponse } from "types/auth/auth.dto.js";
import type { LoginInput } from "types/auth/auth.validation.js";

import * as cron from "node-cron";

export class AuthService {
  private static cleanupJob: cron.ScheduledTask;
  /**
   * Authenticate user and create session
   */
  static async login(
    loginData: LoginInput,
    ipAddress?: string,
    userAgent?: string,
  ): Promise<LoginResponse> {
    const { login_email, password } = loginData;
    const user = await prisma.users.findUnique({
      where: { login_email },
      select: {
        id: true,
        first_name: true,
        last_name: true,
        login_email: true,
        password_hash: true,
        contact_email: true,
        email_verified: true,
        active: true,
        created_at: true,
      },
    });

    if (!user) {
      throw new ApiError(StatusCodes.UNAUTHORIZED, "Invalid email or password");
    }

    if (!user.active) {
      throw new ApiError(StatusCodes.UNAUTHORIZED, "Account is deactivated");
    }

    if (!user.password_hash) {
      throw new ApiError(
        StatusCodes.UNAUTHORIZED,
        "No password set for this account",
      );
    }

    const isPasswordValid = await PasswordUtils.verifyPassword(
      password,
      user.password_hash,
    );
    if (!isPasswordValid) {
      throw new ApiError(StatusCodes.UNAUTHORIZED, "Invalid email or password");
    }

    const placeholderToken = "pending_" + Date.now();
    const sessionData: any = {
      user_id: user.id,
      token_hash: JWTUtils.hashToken(placeholderToken),
      expires_at: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000),
    };

    if (ipAddress) sessionData.ip_address = ipAddress;
    if (userAgent) sessionData.user_agent = userAgent;

    const session = await prisma.user_sessions.create({
      data: sessionData,
    });

    const realToken = JWTUtils.generateToken({
      userId: user.id,
      email: user.login_email,
      sessionId: session.id,
    });

    await prisma.user_sessions.update({
      where: { id: session.id },
      data: {
        token_hash: JWTUtils.hashToken(realToken),
        expires_at: JWTUtils.getTokenExpiry(realToken),
      },
    });

    const { password_hash, ...userWithoutPassword } = user;

    return {
      user: userWithoutPassword,
      token: realToken,
      session_id: session.id,
      expires_at: session.expires_at,
      token_type: "Bearer",
    };
  }

  /**
   * Logout user by revoking session
   */
  static async logout(sessionId: number, userId: number): Promise<boolean> {
    const session = await prisma.user_sessions.updateMany({
      where: {
        id: sessionId,
        user_id: userId,
        revoked_at: null,
      },
      data: {
        revoked_at: new Date(),
      },
    });

    return session.count > 0;
  }

  /**
   * Logout all sessions for user
   */
  static async logoutAll(userId: number): Promise<number> {
    const result = await prisma.user_sessions.updateMany({
      where: {
        user_id: userId,
        revoked_at: null,
        expires_at: { gt: new Date() },
      },
      data: {
        revoked_at: new Date(),
      },
    });

    return result.count;
  }

  /**
   * Validate JWT token and return user
   */
  static async validateToken(
    token: string,
  ): Promise<{ user: any; session: any }> {
    try {
      JWTUtils.verifyToken(token);
      const tokenHash = JWTUtils.hashToken(token);

      const session = await prisma.user_sessions.findFirst({
        where: {
          token_hash: tokenHash,
          revoked_at: null,
          expires_at: { gt: new Date() },
        },
        include: {
          users: {
            select: {
              id: true,
              first_name: true,
              last_name: true,
              login_email: true,
              contact_email: true,
              email_verified: true,
              active: true,
            },
          },
        },
      });

      if (!session) {
        throw new ApiError(
          StatusCodes.UNAUTHORIZED,
          "Invalid or expired session",
        );
      }

      if (!session.users.active) {
        throw new ApiError(StatusCodes.UNAUTHORIZED, "Account is deactivated");
      }

      return {
        user: session.users,
        session: {
          id: session.id,
          ip_address: session.ip_address,
          created_at: session.created_at,
          expires_at: session.expires_at,
        },
      };
    } catch (error) {
      if (error instanceof ApiError) {
        throw error;
      }
      throw new ApiError(
        StatusCodes.UNAUTHORIZED,
        "Invalid authentication token",
      );
    }
  }

  /**
   * Get active sessions for user
   */
  static async getUserSessions(userId: number): Promise<any[]> {
    const sessions = await prisma.user_sessions.findMany({
      where: {
        user_id: userId,
        revoked_at: null,
        expires_at: { gt: new Date() },
      },
      select: {
        id: true,
        ip_address: true,
        user_agent: true,
        created_at: true,
        expires_at: true,
      },
      orderBy: {
        created_at: "desc",
      },
    });

    return sessions;
  }

  /**
   * Initialize session cleanup job
   */
  static initializeSessionCleanup(): void {
    this.cleanupJob = cron.schedule(
      "0 2 * * *",
      async () => {
        try {
          console.log("Starting session cleanup job...");
          const deletedCount = await this.cleanupExpiredSessions();
          console.log(
            `Session cleanup completed: ${deletedCount} sessions removed`,
          );
        } catch (error) {
          console.error("Session cleanup failed:", error);
        }
      },
      {},
    );

    this.cleanupJob.stop();
  }

  /**
   * Start the cleanup job
   */
  static startSessionCleanup(): void {
    if (this.cleanupJob) {
      this.cleanupJob.start();
      console.log("Session cleanup job started");
    }
  }

  /**
   * Stop the cleanup job
   */
  static stopSessionCleanup(): void {
    if (this.cleanupJob) {
      this.cleanupJob.stop();
      console.log("Session cleanup job stopped");
    }
  }

  /**
   * Clean up expired sessions
   */
  static async cleanupExpiredSessions(): Promise<number> {
    const result = await prisma.user_sessions.deleteMany({
      where: {
        OR: [{ expires_at: { lt: new Date() } }, { revoked_at: { not: null } }],
      },
    });

    return result.count;
  }

  /**
   * Manual cleanup (for testing or admin purposes)
   */
  static async manualCleanup(): Promise<{ deleted: number }> {
    const deleted = await this.cleanupExpiredSessions();
    return { deleted };
  }
}

File: internal/core/services/auth.service.go
package services

import (
	"context"
	"errors"
	"log"
	"time"

	"github.com/TIA-PARTNERS-GROUP/tia-api/internal/models"
	"github.com/TIA-PARTNERS-GROUP/tia-api/internal/ports"
	"github.com/TIA-PARTNERS-GROUP/tia-api/pkg/utils"
	"gorm.io/gorm"
)

type AuthService struct {
	db                   *gorm.DB
	sessionCleanupTicker *time.Ticker
	quitChan             chan struct{}
}

func NewAuthService(db *gorm.DB) *AuthService {
	return &AuthService{
		db:       db,
		quitChan: make(chan struct{}),
	}
}

func (s *AuthService) Login(ctx context.Context, data ports.LoginInput, ipAddress, userAgent *string) (*ports.LoginResponse, error) {
	var user models.User
	if err := s.db.WithContext(ctx).Where("login_email = ?", data.LoginEmail).First(&user).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, ports.ErrInvalidCredentials
		}
		return nil, ports.ErrDatabase
	}

	if !user.Active {
		return nil, ports.ErrAccountDeactivated
	}

	if user.PasswordHash == nil || *user.PasswordHash == "" {
		return nil, ports.ErrInvalidCredentials
	}

	if err := utils.VerifyPassword(data.Password, *user.PasswordHash); err != nil {
		return nil, ports.ErrInvalidCredentials
	}

	session := models.UserSession{
		UserID:    user.ID,
		ExpiresAt: time.Now().Add(7 * 24 * time.Hour), // Set a preliminary expiry
	}
	if ipAddress != nil {
		session.IPAddress = ipAddress
	}
	if userAgent != nil {
		session.UserAgent = userAgent
	}
	if err := s.db.WithContext(ctx).Create(&session).Error; err != nil {
		return nil, ports.ErrDatabase
	}

	token, expiry, err := utils.GenerateToken(user.ID, session.ID, user.LoginEmail)
	if err != nil {
		return nil, ports.ErrTokenGeneration
	}

	sessionUpdate := map[string]interface{}{
		"token_hash": utils.HashToken(token),
		"expires_at": expiry,
	}
	if err := s.db.WithContext(ctx).Model(&session).Updates(sessionUpdate).Error; err != nil {
		s.db.Delete(&session)
		return nil, ports.ErrDatabase
	}

	response := &ports.LoginResponse{
		User:      ports.MapUserToResponse(&user),
		Token:     token,
		SessionID: session.ID,
		ExpiresAt: expiry,
		TokenType: "Bearer",
	}

	return response, nil
}

func (s *AuthService) Logout(ctx context.Context, sessionID, userID uint) (bool, error) {
	result := s.db.WithContext(ctx).
		Model(&models.UserSession{}).
		Where("id = ? AND user_id = ? AND revoked_at IS NULL", sessionID, userID).
		Update("revoked_at", time.Now())

	if result.Error != nil {
		return false, ports.ErrDatabase
	}

	return result.RowsAffected > 0, nil
}

func (s *AuthService) LogoutAll(ctx context.Context, userID uint) (int64, error) {
	result := s.db.WithContext(ctx).
		Model(&models.UserSession{}).
		Where("user_id = ? AND revoked_at IS NULL", userID).
		Update("revoked_at", time.Now())

	if result.Error != nil {
		return 0, ports.ErrDatabase
	}
	return result.RowsAffected, nil
}

func (s *AuthService) ValidateToken(ctx context.Context, token string) (*models.User, *models.UserSession, error) {
	claims, err := utils.VerifyToken(token)
	if err != nil {
		return nil, nil, ports.ErrInvalidToken
	}

	var session models.UserSession
	err = s.db.WithContext(ctx).
		Preload("User").
		Where("id = ? AND user_id = ? AND revoked_at IS NULL", claims.SessionID, claims.UserID).
		First(&session).Error

	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil, ports.ErrInvalidSession
		}
		return nil, nil, ports.ErrDatabase
	}

	if time.Now().After(session.ExpiresAt) {
		return nil, nil, ports.ErrInvalidSession
	}

	if !session.User.Active {
		return nil, nil, ports.ErrAccountDeactivated
	}

	return &session.User, &session, nil
}

func (s *AuthService) GetUserSessions(ctx context.Context, userID uint) ([]models.UserSession, error) {
	var sessions []models.UserSession
	err := s.db.WithContext(ctx).
		Where("user_id = ? AND revoked_at IS NULL AND expires_at > ?", userID, time.Now()).
		Order("created_at desc").
		Find(&sessions).Error

	if err != nil {
		return nil, ports.ErrDatabase
	}
	return sessions, nil
}

func (s *AuthService) StartSessionCleanup() {
	log.Println("Initializing session cleanup job...")
	go s.cleanupExpiredSessions()

	s.sessionCleanupTicker = time.NewTicker(24 * time.Hour)
	go func() {
		for {
			select {
			case <-s.sessionCleanupTicker.C:
				log.Println("Running scheduled session cleanup job...")
				s.cleanupExpiredSessions()
			case <-s.quitChan:
				s.sessionCleanupTicker.Stop()
				return
			}
		}
	}()
	log.Println("Session cleanup job started.")
}

func (s *AuthService) StopSessionCleanup() {
	close(s.quitChan)
	log.Println("Session cleanup job stopped.")
}

func (s *AuthService) cleanupExpiredSessions() {
	result := s.db.
		Where("expires_at < ? OR revoked_at IS NOT NULL", time.Now()).
		Delete(&models.UserSession{})

	if result.Error != nil {
		log.Printf("Session cleanup failed: %v", result.Error)
	} else if result.RowsAffected > 0 {
		log.Printf("Session cleanup completed: %d sessions removed", result.RowsAffected)
	}
}
